opsdroid/database/matrix/__init__.py

```diff
@@ -0,0 +1,249 @@
+"""Database that uses the matrix connector."""
+
+import logging
+from contextlib import contextmanager
+
+from nio import RoomGetStateError, RoomGetStateEventError
+from opsdroid.database import Database
+from opsdroid.connector.matrix.events import MatrixStateEvent, GenericMatrixRoomEvent
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DatabaseMatrix(Database):
+    """A module for opsdroid to allow memory to persist in matrix room state."""
+
+    def __init__(self, config, opsdroid=None):
+        """Start the database connection."""
+        super().__init__(config, opsdroid=opsdroid)
+        self.name = "matrix"
+        self.room = config.get("default_room", "main")
+        self._single_state_key = config.get("single_state_key", True)
+        self.should_encrypt = config.get("should_encrypt", False)
+        self._event_type = "dev.opsdroid.database"
+        self.should_migrate = True
+
+        _LOGGER.debug("Loaded matrix database connector.")
+
+    @property
+    def connector(self):
+        """Define the initiated matrix connector."""
+        return self.opsdroid.get_connector("matrix")
+
+    @property
+    def room_id(self):
+        """Define ID of the room to store data in."""
+        return (
+            self.room
+            if self.room.startswith("!")
+            else self.connector.room_ids[self.room]
+        )
+
+    async def migrate_database(self):
+        """Migrate existing 'opsdroid.database' state event to 'dev.opsdroid.database' event."""
+
+        data = await self.connector.connection.room_get_state(room_id=self.room_id)
+        if isinstance(data, RoomGetStateError):
+            _LOGGER.error(
+                f"Error migrating from opsdroid.database to {self._event_type} in room {self.room_id}: {data.message}({data.status_code})"
+            )
+            return
+
+        if not any(event["type"] == self._event_type for event in data.events):
+            for event in data.events:
+                if event["type"] == "opsdroid.database":
```

Comment:
We are going to get here at every startup of the database right, just that the content of the state event (once it's been replaced) would be empty. I think you need this to be:

```suggestion
                if event["type"] == "opsdroid.database" and event["content"]:
```

You should also add a test that the new state event and the redaction isn't sent on a second run of this method.